### PR TITLE
Fix liveness analysis when dealing with phi nodes.

### PR DESCRIPTION
**Pull request by ptersilie**

### CODE REVIEW

**Comment by ltratt** [comment]
FWIW, #191 contains the first commit (plus some other stuff). I suspect git will automerge this without complaint, but it might be worth removing the commit if/when #191 merges?

**Comment by vext01** [comment]
Yeah sorry, I already applied the first commit earlier :)

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp [hunk 1]
```diff
@@ -79,17 +79,23 @@ LivenessAnalysis::LivenessAnalysis(Function *Func) {
...
-      // values being killed immediately after entering the block in the
-      // backwards pass and thus the value never being live there.
+      //   phi [%1, bb1], [%2, bb2], [%3, bb3]
+      //
+      // Here, when the previous block was bb1, then only %1 is used, %2 if we
+      // come from bb2, and so on.
+      //
+      // The next problem is that we can't mark the operand as used in the
+      // current block, since that use would perculate upwards when looking at
+      // the predecessor blocks, and thus consider the operand used in blocks
```
Comment (vext01) [inline llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:94]: > consider the operand used

I'd take the one step further and talk about the value propagating into branches that never defined the variable, thus in the backwards analysis it never has a chance to be killed in those branches, leading to them propagating upwards forever.

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp [hunk 2]
```diff
@@ -79,17 +79,23 @@ LivenessAnalysis::LivenessAnalysis(Function *Func) {
...
+      // where it isn't. Instead, we want to mark the operand as used
+      // only in the predecssor block: %1 is marked as used at the end of bb1,
+      // %2 at the end of bb2, and so on.
```
Comment (vext01) [inline llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:97]: Let's credit Chris Lattner for this simple fix and link to the mailing list post?

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp [hunk 3]
```diff
@@ -79,17 +79,23 @@ LivenessAnalysis::LivenessAnalysis(Function *Func) {
...
       //
-      // The book doesn't cover this quirk, as it explains liveness for
```
Comment (vext01) [inline llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:91]: I think this should stay in.

**Comment by vext01** [comment]
A few comments. Trying to ensure that my future self will be able to easily understand this next time I read it :)

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 2 above)
Comment (ptersilie) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:97]: I thought about it, but I don't think it adds anything, and I don't like reddit links in comments.

Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:97]: Not even the mailing list link?

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 1 above)
Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:94]: I'll do a follow up PR for this.

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 2 above)
Comment (ptersilie) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:97]: The important info in that link is just a single sentence: "PHI uses occur in the predecessor blocks, not in the PHI node blocks".  The rest doesn't seem too relevant: https://lists.llvm.org/pipermail/llvm-dev/2003-November/000542.html

Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:97]: OK, I see you feel strongly, so as you wish.

**Comment by ptersilie** [comment]
Ok to squash?

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed and rebased.

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 3 above)
Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:91]: thanks

**Pull request merged**